File: 3310.remove-methods-from-project.cpp
#
# @lc app=leetcode id=3310 lang=cpp
#
# [3310] Remove Methods From Project
#
# @lc code=start
class Solution {
public:
    vector<int> remainingMethods(int n, int k, vector<vector<int>>& invocations) {
        // Initialize adjacency lists for direct and reverse graphs
        vector<vector<int>> adjList(n), reverseAdjList(n);
        for (auto& invocation : invocations) {
            adjList[invocation[0]].push_back(invocation[1]);
            reverseAdjList[invocation[1]].push_back(invocation[0]);
        }
        
        // Perform BFS/DFS from method k to find all reachable/suspicious nodes
        vector<bool> isSuspicious(n, false);
        queue<int> q;
        q.push(k);
        isSuspicious[k] = true;
        while (!q.empty()) {
            int current = q.front(); q.pop();
            for (int neighbor : adjList[current]) {
                if (!isSuspicious[neighbor]) {
                    isSuspicious[neighbor] = true;
                    q.push(neighbor);
                }
            }
        }

        // Check if any non-suspicious method invokes a suspicious one
        for (int i = 0; i < n; ++i) {
            if (!isSuspicious[i]) {
                for (int neighbor : reverseAdjList[i]) {
                    if (isSuspicious[neighbor]) {
                        // If any such invocation exists, we cannot safely remove the suspicious group.
                        return vector<int>(begin({0}), end({0}) + n); // return all methods
                    }
                }
            }
        }

        // Collect non-suspicious methods
        vector<int> remainingMethods;
        for (int i = 0; i < n; ++i) {
            if (!isSuspicious[i]) {
                remainingMethods.push_back(i);
            }
        }

        return remainingMethods;
    }
};
# @lc code=end

File: 3310.remove-methods-from-project.java
#
# @lc app=leetcode id=3310 lang=java
#
# [3310] Remove Methods From Project
#

# @lc code=start
class Solution {
    public List<Integer> remainingMethods(int n, int k, int[][] invocations) {
        // Step 1: Build graph from invocations
        List<Integer>[] graph = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
        }
        for (int[] edge : invocations) {
            graph[edge[0]].add(edge[1]);
        }
        
        // Step 2: Find all suspicious methods starting from k using DFS
        Set<Integer> suspicious = new HashSet<>();
        dfs(k, graph, suspicious);
        
        // Step 3: Check if any method outside suspicious invokes any within it
        boolean[] invokedByOutside = new boolean[n];
        for (int[] edge : invocations) {
            if (!suspicious.contains(edge[0]) && suspicious.contains(edge[1])) {
                invokedByOutside[edge[1]] = true;
            }
        }
        for (int method : suspicious) {
            if (invokedByOutside[method]) {
                return IntStream.range(0, n).boxed().collect(Collectors.toList()); // Return all methods if condition not met. 
            } 
        } 
        
        // Step 4: Return remaining methods after removing suspicious ones if possible. 
        List<Integer> remainingMethods = new ArrayList<>(); 
        for (int i = 0; i < n; i++) { 
            if (!suspicious.contains(i)) { 
                remainingMethods.add(i); 
            } 
        }  return remainingMethods;  } private void dfs(int node, List<Integer>[] graph, Set<Integer> visited) { if (visited.contains(node)) return; visited.add(node); for (int neighbor : graph[node]) dfs(neighbor, graph, visited); } # @lc code=end